### PR TITLE
[web] - sync text editing scroll state, pass `line-height` and `letter-spacing` to engine

### PR DESCRIPTION
**Pull request by htoor3**
We need to send scroll state, line-height, and letter-spacing to the engine to more closely align the web text inputs overlays. 

Framework changes to support https://github.com/flutter/flutter/issues/135159

Engine changes are here https://github.com/flutter/engine/pull/46109

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 1]
```diff
@@ -2917,6 +2918,8 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
           fontWeight: _style.fontWeight,
           textDirection: _textDirection,
           textAlign: widget.textAlign,
+          lineHeight: _style.height,
+          letterSpacing: _style.letterSpacing,
```
Comment (htoor3) [inline packages/flutter/lib/src/widgets/editable_text.dart:2922]: Do these translate directly to their CSS equivalents? In my basic testing, it seems to work out fine - however, our y-translation from the transforms we apply end up being off by a couple of pixels

**Comment by Renzo-Olivares** [comment]
@htoor3 Passing `line-height` and `letter-spacing` makes sense to me. From the documentation it seems like `TextStyle.letterSpacing` translates directly to its CSS equivalent `letter-spacing`. However, `height` seems to be different then the CSS equivalent of `line-height`.

`TextStyle.height` in flutter is a multiplier for the line height defined by the font metrics. When it is `null` the line height is defined completely by the font metrics. 

I think the `line-height` value we are looking for might be https://api.flutter.dev/flutter/dart-ui/LineMetrics/height.html.

Something else to consider but probably out of scope of this pr is lines with multiple styles.

WRT to the scroll state I can't think of any immediate issues with this approach but I think there may be issues if the `EditableText` is within a `Scrollable`. For example this issue https://github.com/flutter/flutter/issues/129590 . 

cc @LongCatIsLooong for any thoughts regarding these changes.

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 1 above)
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/editable_text.dart:2922]: Replied to this in https://github.com/flutter/flutter/pull/135163#issuecomment-1758654663

**Comment by LongCatIsLooong** [comment]
Is this for the HTML renderer or canvaskit/skwasm?

`line-height`: I think `TextStyle.lineHeight` is interpreted as a percentage: 1.0 => 100% of `TextStyle.fontSize` (in logical pixels). E.g.: 14 px fontSize, 1.5 line height => about 21 px line height. But there are a couple of catches. If the glyphs in a line use different fonts, then the actual line height is going to be taller. Additionally, when `TextStyle.lineHeight` is not 1.0 or null, by default in `SkParagraph`/`CanvasKit` the baseline position could be a bit different from that in browsers (see: https://master-api.flutter.dev/flutter/painting/TextStyle-class.html#painting.TextStyle.4).

For `letter-spacing` afaik SkParagraph divides the value by 2 and increases each glyph clusters' advance by that much.

**Comment by LongCatIsLooong** [comment]
Since this is web only these messages should probably only be sent when `kIsWeb` is true. Also is it possible to send the scroll offsets only at the moment the scroll view stops scrolling? That's going to reduce the number of messages by a lot.

**Comment by goderbauer** [comment]
(Triage): @htoor3 Do you still have plans for this PR?

**Comment by htoor3** [comment]
> (Triage): @htoor3 Do you still have plans for this PR?

We've lowered the priority of https://github.com/flutter/flutter/issues/135159 to focus on other higher priority issues - would you prefer I close this in the meantime?

**Comment by htoor3** [comment]
Thank you for the comments @Renzo-Olivares & @LongCatIsLooong! A good action item when I return to this is to debounce the platform messages for scroll events & only send them when `kIsWeb` is `true` (which is currently happening on framework scroll events, but I should also add it on opening a new connection). 

From the comments, it seems like we can rely on `letter-spacing` but we probably need to be a little more careful with `TextStyle.height`.

> line-height: I think TextStyle.lineHeight is interpreted as a percentage: 1.0 => 100% of TextStyle.fontSize (in logical pixels). E.g.: 14 px fontSize, 1.5 line height => about 21 px line height. But there are a couple of catches. If the glyphs in a line use different fonts, then the actual line height is going to be taller. Additionally, when TextStyle.lineHeight is not 1.0 or null, by default in SkParagraph/CanvasKit the baseline position could be a bit different from that in browsers (see: https://master-api.flutter.dev/flutter/painting/TextStyle-class.html#painting.TextStyle.4).

Based on [this](https://developer.mozilla.org/en-US/docs/Web/CSS/line-height), the browser also interprets the (unitless) value as a percentage multiplier for the set font-size, so I think this could also work well if we're careful about the edge cases you've mentioned.  

We've definitely seen some rendering inconsistency issues especially with mismatched fonts and have some todo items (see https://github.com/flutter/flutter/issues/135159). We also don't have the same degree of granularity when rendering text on the browser, as I don't think there's a way for us change the baseline position via CSS like we can in Flutter.  Although getting the texts to line up exactly would be ideal, I think we can also accept a "close enough" approximation as this would help solve some of the issues we're seeing like https://github.com/flutter/flutter/issues/134024.

**Comment by goderbauer** [comment]
> We've lowered the priority of https://github.com/flutter/flutter/issues/135159 to focus on other higher priority issues - would you prefer I close this in the meantime?

If no action is going to be taken on this PR in the near future we should just close it. Feel free to re-open when more work is scheduled for it.

**Pull request closed without merging**